Fix for issue 40774 RequestDelegateFactory null param default

Acerby opened the pull request:
# Fix for issue 40774 RequestDelegateFactory null param default

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

Fix issue with default value for nullable struct parameters

Added missing parameter type info for expression tree constant

Also added test to verify this functionality

Fixes #40774  (in this specific format)

Acerby committed:
Added test for nullable structs

Fixed typo in test RequestDelegatePopulatesFromNullableOptionalParameter that was previously just a duplicate of the previous test.
Plus added a new test case for handling a default (null) value as a parameter (Test for Issue dotnet#40774)

Acerby committed:
Fix for issue 40774

Acerby committed:
Merge branch 'dotnet:main' into main

Acerby committed:
Asset.Equals(null,... to Assert.IsNull(

campersau commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -387,7 +387,7 @@ public async Task RequestDelegatePopulatesFromNullableNullOptionalParameter()
         await requestDelegate(httpContext);
 
-        Assert.Equal(null, httpContext.Items["input"]);
+        Assert.IsNull(httpContext.Items["input"]);
```
```suggestion
        Assert.Null(httpContext.Items["input"]);
```

Acerby committed:
Assert.IsNull( to Assert.Null(

Thanks @campersau

halter73 commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -226,6 +226,11 @@ private static void TestOptionalNullable(HttpContext httpContext, int? value = 4
 
+    private static void TestOptionalNullableNull(HttpContext httpContext, double? value = null)
+    {
+        httpContext.Items.Add("input", value);
```
```suggestion
        httpContext.Items.Add("input", value ?? "null");
```

The `ItemsDictionary` is unusual in that it returns `null` for missing values, so it's impossible to know from a test if this line executed or not.

https://github.com/dotnet/aspnetcore/blob/e001dc01a152806a457eb9da15dc102107451dee/src/Http/Http/src/Internal/ItemsDictionary.cs#L23-L33

Acerby replied:
Many thanks for spotting that @halter73. I have updated the Acerby:main branch based on a slightly updated version your suggestion of using "null".

Acerby committed:
Improved test to verify method actually ran

As kindly pointed out by  @halter73 the httpContext.Items[index] return null be default so using explicit "Null" value to verify method was actually run

Acerby committed:
Merge branch 'dotnet:main' into main

Acerby committed:
Patch for missing change/merge issue

Acerby committed:
Reverting unintentional? change

halter73 commented:
@Acerby thanks for both finding and fixing this issue!
